// @toio/cube/lib/characteristics/configuration-characteristic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var ConfigurationCharacteristic = (function () {
    function ConfigurationCharacteristic(characteristic) {
        this.eventEmitter = new events_1.EventEmitter();
        this.characteristic = characteristic;
        if (this.characteristic.properties.includes('notify')) {
            this.characteristic.on('data', this.onData.bind(this));
        }
    }
    ConfigurationCharacteristic.prototype.getBLEProtocolVersion = function () {
        var _this = this;
        return new Promise(function (resolve) {
            _this.characteristic.subscribe();
            _this.characteristic.write(Buffer.from([0x01, 0x00]), false);
            _this.eventEmitter.once('configuration:ble-protocol-version', function (version) {
                _this.characteristic.unsubscribe();
                resolve(version);
            });
        });
    };
    ConfigurationCharacteristic.prototype.data2result = function (data) {
        var type = data.readUInt8(0);
        if (type === 0x81) {
            var version = data.toString('utf-8', 2, 7);
            this.eventEmitter.emit('configuration:ble-protocol-version', version);
            return;
        }
    };
    ConfigurationCharacteristic.prototype.onData = function (data) {
        this.data2result(data);
    };
    ConfigurationCharacteristic.UUID = '10b201ff5b3b45719508cf3efcd7bbae';
    return ConfigurationCharacteristic;
}());
exports.ConfigurationCharacteristic = ConfigurationCharacteristic;
