// @toio/cube/lib/characteristics/motor-characteristic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var motor_spec_1 = require("./specs/motor-spec");
var MotorCharacteristic = (function () {
    function MotorCharacteristic(characteristic) {
        this.spec = new motor_spec_1.MotorSpec();
        this.timer = null;
        this.pendingResolve = null;
        this.characteristic = characteristic;
    }
    MotorCharacteristic.prototype.move = function (left, right, durationMs) {
        var _this = this;
        if (this.timer) {
            clearTimeout(this.timer);
            this.timer = null;
        }
        if (this.pendingResolve) {
            this.pendingResolve();
            this.pendingResolve = null;
        }
        var data = this.spec.move(left, right, durationMs);
        this.characteristic.write(Buffer.from(data.buffer), false);
        if (data.data.durationMs > 0) {
            return new Promise(function (resolve) {
                _this.pendingResolve = resolve;
                _this.timer = setTimeout(function () {
                    if (_this.pendingResolve) {
                        _this.pendingResolve();
                        _this.pendingResolve = null;
                    }
                }, data.data.durationMs);
            });
        }
    };
    MotorCharacteristic.prototype.stop = function () {
        this.move(0, 0, 0);
    };
    MotorCharacteristic.UUID = '10b201025b3b45719508cf3efcd7bbae';
    return MotorCharacteristic;
}());
exports.MotorCharacteristic = MotorCharacteristic;
